File: Leetcode/Binary Tree Postorder Traversal.cc
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> postorderTraversal(TreeNode *root) {
        stack<TreeNode*> st;
        vector<int> res;
        
        if (!root) return res;
        
        TreeNode *pre = NULL;
        st.push(root);
        
        while (!st.empty()) {
            TreeNode *cur = st.top();
            if ( (!cur -> left && !cur -> right) ||
                    (pre && (pre == cur -> left || pre == cur -> right))) {
                res.push_back(cur -> val);
                st.pop();
                pre = cur;
            } else {
                if (cur -> right) st.push(cur -> right);
                if (cur -> left) st.push(cur -> left);
            }
        }
        return res;
    }
};

File: Leetcode/python/Binary Tree Postorder Traversal.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a list of integers
    def postorderTraversal(self, root):
        if not root:
            return []
        st, res = [root], []
        prev = None
        
        while st:
            cur = st[-1]
            if (not cur.left and not cur.right) or \
                    (prev and (prev == cur.left or prev == cur.right)):
                res.append(cur)
                st.pop()
                prev = cur
            else:
                if cur.right:
                    st.append(cur.right)
                if cur.left:
                    st.append(cur.left)
        return map(lambda x: x.val, res)
